### ISSUE TITLE
biopar: speed up computation?

### ISSUE THREAD

**Issue by jdries** [Open-EO/openeo-geopyspark-driver#682]
This code is apparently relatively slow:
https://git.vito.be/projects/GEOM/repos/biopar/browse/src/biopar/bioparnnw.py#78

And it logs this:
`11 out of the last 11 calls to <function BioParNNW._compute_biopar at 0x7fad3f8fe4c0> triggered tf.function retracing. Tracing is expensive and the excessive number of tracings could be due to (1) creating @tf.function repeatedly in a loop, (2) passing tensors with different shapes, (3) passing Python objects instead of tensors. For (1), please define your @tf.function outside of the loop. For (2), @tf.function has experimental_relax_shapes=True option that relaxes argument shapes that can avoid unnecessary retracing. For (3), please refer to https://www.tensorflow.org/tutorials/customization/performance#python_or_tensor_args and https://www.tensorflow.org/api_docs/python/tf/function for  more details.`

The UDF that invokes this also does not cache the biopar class.
It's also an option to write a numpy implementation next to the tensorflow one.

Reading the tf.function documentation again, it very much looks like we want to try avoiding the use of a class method, and just replace the variables coming from there with tensorflow constants.

**Comment by jdries** [comment]
The UDF that invokes this:


import numpy as np
from typing import Dict
import xarray as xr 
from openeo.udf.xarraydatacube import XarrayDataCube
import tensorflow as tf
from biopar.bioparnnw import BioParNNW

nbiopar_version = '3band'
def apply_datacube(cube: XarrayDataCube, context: Dict) -> XarrayDataCube:
    valid_biopars= ['FAPAR','LAI','FCOVER','CWC','CCC']
    biopar = context.get('biopar', 'FAPAR')
    if biopar not in valid_biopars:
                biopar = 'FAPAR'
    
    inarr = cube.get_array()
    ds_date = inarr
    
    from numpy import cos, radians
    scaling_bands = 0.0001
    saa = ds_date.sel(bands='sunAzimuthAngles')
    sza = ds_date.sel(bands='sunZenithAngles')
    vaa = ds_date.sel(bands='viewAzimuthMean')
    vza = ds_date.sel(bands='viewZenithMean')
    
    B03 = ds_date.sel(bands='B03') * scaling_bands
    B04 = ds_date.sel(bands='B04') * scaling_bands
    B8 = ds_date.sel(bands='B08') * scaling_bands
    g1 = cos(radians(vza))
    g2 = cos(radians(sza))
    g3 = cos(radians(saa - vaa))
    #### FLATTEN THE ARRAY ####
    flat = list(map(lambda arr: arr.flatten(), [B03.values, B04.values,B8.values, g1.values, g2.values, g3.values]))
    bands = np.array(flat)
    #### CALCULATE THE BIOPAR BASED ON THE BANDS #####
    image = BioParNNW(version='3band', parameter=biopar, singleConfig = True).run(bands, output_scale=1,output_dtype=tf.dtypes.float32,minmax_flagging=False)  # netcdf algorithm
    as_image = image.reshape((g1.shape))
    ## set nodata to nan
    as_image[np.where(np.isnan(B03))] = np.nan
    xr_biopar = vza.copy()
    xr_biopar.values = as_image / scaling_bands
    
    return XarrayDataCube(xr_biopar)  # xarray.DataArray(as_image,vza.dims,vza.coords)

**Comment by VictorVerhaert** [comment]
I have requested access to the "CGS_S2" folder in order to run the 3_band tests and hopefully recreate the warnings.

**Comment by jdries** [comment]
@VictorVerhaert I see, these are in fact old paths that can simply be updated with more recent versions:
/data/MTDA/TERRASCOPE_Sentinel2/TOC_V2/2017/06/02/S2A_20170602T104021_31UFS_TOC_V210
there should be a similar path to a corresponding fapar product, to use as reference

I would however also recommend the 8 band test:
https://git.vito.be/projects/GEOM/repos/biopar/browse/tests/test_8_band.py
which uses the same code, but does not have complex test dependencies

**Comment by VictorVerhaert** [comment]
Pull request can be found [here](https://git.vito.be/projects/GEOM/repos/biopar/pull-requests/2/overview).
For now I have only added a new class so both implementations can be executed and compared in terms of speed, memory usage and results.

Where can I find a script that uses the biopar module? that way I can set up the comparison.
(Using the job_id mentioned before I got a process graph but this graph gives me an error)
